text/3305-static-in-pattern.md

```diff
@@ -0,0 +1,285 @@
+- Feature Name: `static_in_pattern`
+- Start Date: 2022-08-17
+- RFC PR: [rust-lang/rfcs#3305](https://github.com/rust-lang/rfcs/pull/3305)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+[summary]: #summary
+
+Allow referencing non-`mut` `static`s in pattern matches wherever referencing a `const` of the same type would be allowed.
+
+# Motivation
+
+[motivation]: #motivation
+
+Rust pattern matches compare a scrutinee against compile-time information. Rust generally doesn't allow patterns to depend on runtime information; that is relegated to match guards. However, there is a category between "compile-time", when `rustc` runs, and "runtime", when Rust code runs. Some information a Rust program relies on may be determined at link-time, or by the target operating system, or before `main()` by the C runtime. Rust currently prevents patterns from depending on such information. Specifically, Rust patterns cannot reference statics from `extern` blocks.
+
+I encountered this restriction while trying to port the Rust standard library to [cosmopolitan libc](https://justine.lol/cosmopolitan/index.html). Cosmopolitan provides an API that mostly matches POSIX, with one major exception: constants like `ENOSYS` and `EINVAL`, which on most platforms are defined as C `#define`s (equivalent to Rust `const`s), are instead provided as C `const`s (equivalent to Rust non-`mut` `static`s).
```

Comment:
Not only are they defined like that on most platforms, [POSIX actually requires them to be defined like that](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/errno.h.html#tag_13_09_03), as @eddyb pointed out
> The \<errno.h> header shall define the following macros which shall expand to integer constant expressions with type int, distinct positive values (except as noted below), and which shall be suitable for use in #if preprocessing directives:

The RFC should mention that. I am not sure if an unportable POSIX-violating libc is sufficient justification for a language extension -- it would be good to have some other, less niche usecases as well.